### DISCUSSION TITLE
How to use docker-compose with GitHub Actions?

### DISCUSSION THREAD

**Discussion by ethanjahn** [community/community#27185] [Actions]
<p>Hi!</p>
<p>I am trying to use GitHub actions to automate my test pipeline, but cannot seem to get the containers to run in order to test them. I am running a django webapp in one container and postgres in another and tying the two with docker-compose.</p>
<p>My docker-compose.yml file is</p>
<pre><code>version: '3'

services:
web:
container_name: backend
build: .
volumes:
- ~/app_name:/code
ports:
- "8000:8000"
environment:
- PORT=${PORT}
- DJANGO_SETTINGS_MODULE=${DJANGO_SETTINGS_MODULE}
depends_on:
- postgres

postgres:
container_name: postgres
image: postgres
restart: always
environment:
- POSTGRES_USER=${PG_USER}
- POSTGRES_PASSWORD=${PG_PASS}
- POSTGRES_DB=${PG_DB}
ports:
- 5432:5432
volumes:
- pgdata:/var/lib/postgresql/data

volumes:
pgdata:
</code></pre>
<p>and my GitHub actions file is:</p>
<pre><code>name: Docker Image CI

on: [push]

jobs:

build:

runs-on: ubuntu-latest
env:
DATABASE_URL: ${{ secrets.DATABASE_URL }}
DJANGO_SETTINGS_MODULE: app_name.settings.dev
SECRET_KEY: ${{ secrets.SECRET_KEY }}
PORT: 8000
PG_DB: ${{ secrets.PG_DB }}
PG_PASS: ${{ secrets.PG_PASS }}
PG_USER: ${{ secrets.USER }}

steps:
- uses: actions/checkout@v1
- name: Build the docker-compose stack
run: docker-compose up -d
- name: Sleep
uses: jakejarvis/wait-action@master
with:
time: '60s'
- name: Check running containers
run: docker ps
- name: Run test suite
run: docker exec backend pytest --skip-auth
</code></pre>
<p>Note that I added the 60s wait time, because I thought the test suite was running before the container could stand up but it didn’t seem to help.</p>
<p>The output of the “docker ps” call is below. Note that the web service is not running:</p>
<pre><code>CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
d56c857f7316 postgres "docker-entrypoint.s…" About a minute ago Up About a minute 0.0.0.0:5432-&gt;5432/tcp postgres
</code></pre>
<p>and the output of the test step is:</p>
<pre><code>Error response from daemon: Container c9e106fab289d59c8fd233544cc7a642412f315241981bf3e00edba0a6a186eb is not running
##[error]Process completed with exit code 1.
</code></pre>
<p>This docker-compose file works great locally but something seems to be amiss when running on GitHub actions. Really appreciate any advice you have!</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Glad to hear you in GitHub Community.</p>
<p>Please try to run <strong>docker ps -a</strong> It will list all containers including exited ones. Based on your docker-compose.yml, it seems that backend container has exited immediately after building the image.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/df6f81ccd16b1c49531755385b861a14e140de2e.png" data-download-href="/uploads/short-url/vSBntqyc4zlqFERkVgBKArZpQvQ.png?dl=1" title="docker ps.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181102180-eeb80d9b-12cd-4fdb-9cb0-87a20f99a998.png" alt="docker ps.png" data-base62-sha1="vSBntqyc4zlqFERkVgBKArZpQvQ" width="690" height="209" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/df6f81ccd16b1c49531755385b861a14e140de2e_2_690x209.png, https://aws1.discourse-cdn.com/github/optimized/2X/d/df6f81ccd16b1c49531755385b861a14e140de2e_2_1035x313.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/d/df6f81ccd16b1c49531755385b861a14e140de2e.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/df6f81ccd16b1c49531755385b861a14e140de2e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">docker ps.png</span><span class="informations">1150×349 16 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Use jakejarvis/wait-action@master action could not keep container running.</p>
<p>You can try to add next command at the end of your docker file.</p>
<p><strong>CMD tail -f /dev/null</strong></p>
<p>You can refer to this blog.</p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=http%3A%2F%2Fbigdatums.net%2F2017%2F11%2F07%2Fhow-to-keep-docker-containers-running%2F&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7Cd8cfb360682348ce762f08d767455af4%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637091419884290226&amp;sdata=r7gg%2BfvAIql%2BsL9BwFvE%2FGWAvCTy3pKh%2BbV7tILl3SE%3D&amp;reserved=0" rel="nofollow noopener">http://bigdatums.net/2017/11/07/how-to-keep-docker-containers-running/</a></p>

**Comment by ethanjahn** [comment]
<p>Hi Yanjingzhu,</p>
<p>Thanks for your help! It turns out the issue was actually with my volume command in the docker-compose file, which was looking for a volume that doesn’t exist on the GitHub Actions VM.</p>
<p>To debug, I used the command:</p>
<pre><code>docker logs backend
</code></pre>
<p>Which pointed out that gunicorn was unable to find my django app:</p>
<pre><code>Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/workers/base.py", line 129, in init_process
    self.load_wsgi()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
    return self.load_wsgiapp()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/usr/local/lib/python3.7/site-packages/gunicorn/util.py", line 350, in import_app
    __import__ (module)
ModuleNotFoundError: No module named 'app_name'
</code></pre>
<p>I think that the volume I was calling was actually overwriting all my code, since there was nothing in the volume in the actions VM. I didn’t test this thouroughly, but I did get it to work by splitting my docker-compose file into three files: A base configuration, a local configuration, and a ci configuration</p>
<p>Base config</p>
<pre><code># docker-compose.yml
version: '3'

services:
  web:
    container_name: backend
    build: .
    ports:
      - "${PORT}:8000"
    depends_on:
      - postgres

  postgres:
    container_name: postgres
    image: postgres
    restart: always
    environment:
      - POSTGRES_PASSWORD=${PG_PASS}
    ports:
      - 5432:5432
</code></pre>
<p>Local dev config:</p>
<pre><code># docker-compose.override.yml
version: '3'

services:
  web:
    env_file:
      - .env

# This was the offending line
    volumes:
      - ~/app_name/:/code

  postgres:
      volumes:
      - pgdata:/var/lib/postgresql/data

volumes:
  pgdata:
</code></pre>
<p>CI config</p>
<pre><code># docker-compose.ci.yml
version: '3'

services:
  web:
    environment:
      - DATABASE_URL
      - DJANGO_SETTINGS_MODULE
      - SECRET_KEY
      - PORT
</code></pre>
<p>I then updated my github actions yml file:</p>
<pre><code>name: Docker Image CI

on: [push]

jobs:

  build:

    runs-on: ubuntu-latest
    env:
      PG_PASS: ${{ secrets.PG_PASS }}
      DATABASE_URL: postgresql://postgres:${{ secrets.SECRET_PASSWORD }}@postgres/postgres
      DJANGO_SETTINGS_MODULE: app_name.settings.dev
      SECRET_KEY: ${{ secrets.SECRET_KEY }}
      PORT: 8000

    steps:
    - uses: actions/checkout@v1
    - name: Build the docker-compose stack
      run: docker-compose -f docker-compose.yml -f docker-compose.ci.yml up -d
    - name: Check running containers
      run: docker ps -a
    - name: Check logs
      run: docker logs backend
    - name: Run test suite
      run: docker exec backend pytest --skip-auth
</code></pre>
<p>The main change to the GitHub actions file is changing the docker-compose command to use the docker-compose.ci.yml file.</p>
<p>Hope this is helpful to future users! Looks like docker-compose works as expected, I just forgot that my local volume would not be available on the server.</p>

**Comment by espoirMur** [comment]
<p>Man, You saved my life!</p>
<p>The docker volumes thing gaves me headaches…</p>

**Comment by alanwilter** [comment]
<p>Hi, I look at your issue and solution, I’m sure my case is very similar but I’m not working out a solution. The error is almost identical, yet…</p>
<p>I’m using <a href="https://github.com/nektos/act" rel="nofollow noopener"><code>act</code></a> with <a href="https://hub.docker.com/r/nektos/act-environments-ubuntu/tags" rel="nofollow noopener">nektos/act-environments-ubuntu:18.04</a> full 18GB docker image to debug.</p>
<p>Here is <code>docker-compose.yml</code>:</p>
<pre><code class="lang-auto">version: '3'
services:
   app:
      build:
         context: .
      image: my_linux_python
      command: sh -c "gunicorn -b 0.0.0.0:5000 --reload --workers=1 --threads=15 application:application"
      ports:
      - 5000:5000
      working_dir: /app
      volumes:
      - ./:/app
      env_file:
      - public.env
      #- private.env
      depends_on:
      - db
   db:
      image: postgres:12-alpine
      volumes:
      - db:/var/lib/postgresql/data
      - ./:/app
      - ./db/import_schema.sql:/docker-entrypoint-initdb.d/1_import_schema.sql
      - ./db/import_demo_data.sql:/docker-entrypoint-initdb.d/2_import_demo_data.sql
      env_file:
      - public.env
      working_dir: /app
volumes:
   db:
</code></pre>
<p>Just in case it helps, my <code>Dockerfile</code>, used to create <code>my_linux_python</code> image:</p>
<pre><code class="lang-bash">FROM debian:buster-slim

# set work directory
WORKDIR /app

# set environment variables, to avoid pyc files and flushing buffer
ENV PYTHONDONTWRITEBYTECODE 1
ENV PYTHONUNBUFFERED 1

COPY ./requirements.txt /app/requirements.txt

RUN apt-get update \
    &amp;&amp; apt-get install --no-install-recommends -y python3-pip=18.1-5 python3-pysam=0.15.2+ds-2 \
    &amp;&amp; pip3 --no-cache-dir install --upgrade pip \
    &amp;&amp; pip --no-cache-dir install setuptools==49.1.0 gunicorn==20.0.4 \
    &amp;&amp; pip --no-cache-dir install -r requirements.txt \
    &amp;&amp; apt-get autoremove -y &amp;&amp; apt-get autoclean -y &amp;&amp; apt-get clean -y \
    &amp;&amp; rm -rf /var/lib/apt/lists/*
</code></pre>
<p>I’m trying to run my <code>docker-compose up</code> inside <code>act</code> but it fails with the same error you’ve seen.</p>
<pre><code class="lang-auto">[2020-08-10 13:21:25 +0000] [6] [INFO] Starting gunicorn 20.0.4
[2020-08-10 13:21:25 +0000] [6] [INFO] Listening at: http://0.0.0.0:5000 (6)
[2020-08-10 13:21:25 +0000] [6] [INFO] Using worker: threads
[2020-08-10 13:21:25 +0000] [9] [INFO] Booting worker with pid: 9
[2020-08-10 13:21:25 +0000] [9] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/workers/gthread.py", line 92, in init_process
    super().init_process()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/workers/base.py", line 119, in init_process
    self.load_wsgi()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/workers/base.py", line 144, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/app/wsgiapp.py", line 49, in load
    return self.load_wsgiapp()
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/app/wsgiapp.py", line 39, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/usr/local/lib/python3.7/dist-packages/gunicorn/util.py", line 358, in import_app
    mod = importlib.import_module(module)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "&lt;frozen importlib._bootstrap&gt;", line 1006, in _gcd_import
  File "&lt;frozen importlib._bootstrap&gt;", line 983, in _find_and_load
  File "&lt;frozen importlib._bootstrap&gt;", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'application'
[2020-08-10 13:21:25 +0000] [9] [INFO] Worker exiting (pid: 9)
[2020-08-10 13:21:26 +0000] [6] [INFO] Shutting down: Master
[2020-08-10 13:21:26 +0000] [6] [INFO] Reason: Worker failed to boot.
</code></pre>
<p>My current <code>python-app.yml</code>, for debugging is:</p>
<pre><code class="lang-auto">name: Python application
on:
   push:
      branches:
      - prod-live
   pull_request:
      branches:
      - prod-live
jobs:
   build:
      runs-on: ubuntu-latest
      steps:
      -  uses: actions/checkout@v2
      -  name: Run docker-compose stack
         run: docker-compose -f docker-compose.yml up -d
      -  name: Check folder
         run: pwd
      -  name: Check files
         run: ls -ltr
# if docker container app were running then... 
      -  name: Lint with flake8
         run: docker-compose exec app flake8
      -  name: Check format with black
         run: docker-compose exec app black --diff --check .
      -  name: Test with PyTest
         run: docker-compose exec app python3 -m pytest 
</code></pre>
<p>Then, when I run <code>act</code>:</p>
<pre><code class="lang-auto">[Python application/build] 🚀  Start image=nektos/act-environments-ubuntu:18.04
[Python application/build]   🐳  docker run image=nektos/act-environments-ubuntu:18.04 entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Python application/build]   🐳  docker cp src=/Users/alan/Programmes/phenopolis_api/. dst=/github/workspace
[Python application/build] ⭐  Run actions/checkout@v2
[Python application/build]   ✅  Success - actions/checkout@v2
[Python application/build] ⭐  Run Run docker-compose stack
Starting workspace_db_1 ... done
Starting workspace_app_1 ...
[Python application/build]   ✅  Success - Run docker-compose stack
Starting workspace_app_1 ... done
/github/workspace
[Python application/build]   ✅  Success - Check folder
[Python application/build] ⭐  Run Check files
| total 84
| -rw-r--r-- 1  502 dialout    84 Jun 25 12:46 Procfile
| -rw-r--r-- 1  502 dialout   470 Jul  5 16:06 env_vars.sh
| -rw-r--r-- 1  502 dialout  2306 Jul  8 21:22 README.md
| -rw-r--r-- 1  502 dialout 13206 Jul  8 21:26 code_setup.md
| -rw-r--r-- 1  502 dialout    31 Jul 27 09:49 pyproject.toml
| -rw-r--r-- 1  502 dialout   128 Jul 27 09:49 application.py
| -rw-r--r-- 1  502 dialout   105 Aug  3 21:50 public.env
| -rw-r--r-- 1  502 dialout   636 Aug 10 07:48 Dockerfile
| -rw-r--r-- 1  502 dialout   377 Aug 10 08:10 requirements.txt
| -rw-r--r-- 1  502 dialout   612 Aug 10 12:22 tox.ini
| -rw-r--r-- 1  502 dialout   735 Aug 10 12:52 docker-compose.yml
| -rw-r--r-- 1  502 dialout   739 Aug 10 12:59 docker-compose2.yml
| drwxr-xr-x 3 root root     4096 Aug 10 14:13 db
| drwxr-xr-x 2 root root     4096 Aug 10 14:13 response_templates
| drwxr-xr-x 4 root root     4096 Aug 10 14:13 schema
| drwxr-xr-x 2 root root     4096 Aug 10 14:13 scripts
| drwxr-xr-x 2 root root     4096 Aug 10 14:13 tests
| drwxr-xr-x 2 root root     4096 Aug 10 14:13 views
[Python application/build]   ✅  Success - Check files
[Python application/build] ⭐  Run Lint with flake8
| ERROR: No container found for app_1
[Python application/build]   ❌  Failure - Lint with flake8
Error: exit with `FAILURE`: 1
</code></pre>
<p>I understand that volume containing the app is not well setup, but how to change that? In your example <code>docker-compose.ci.yml</code> and <code>docker-compose.yml</code> don’t mention volumes, so how can it work?<br>
Sorry if I missed something, I’m new to GitHup action as well.</p>

**Comment by Raghav2211** [comment]
<p>Hi, I’ve user curl command to test spring-boot docker deployment like below:</p>
<pre><code class="lang-auto">    - name: Build with Maven
      run: mvn -B package
    - name: Build local image
      run: docker build -f Dockerfile.mvn --tag &lt;name&gt;:&lt;tag&gt;.
    - name: Build the stack
      run: docker-compose --env-file=env/local/Docker.env up -d
    - name: Test
      run: |
        sleep 120
        docker run --network container:&lt;container-name&gt; appropriate/curl -s --retry 10 --retry-connrefused http://localhost:8080/actuator/health
</code></pre>

**Reply by denzuko** [reply]
Does not add to the convo nor answer OP's question.

**Comment by iamliamc** [comment]
I have a question is using docker-compose here the best practice or would it be a considerable improvement to use: 
https://docs.github.com/en/actions/use-cases-and-examples/using-containerized-services/about-service-containers

**Reply by lgrosz** [reply]
I think this depends on what you need. If you can get away with using a public/in-registry image then service containers seem much cleaner.. but if you need to build an image as a part of the job then service containers cannot be used since they are loaded before the jobs start.
